File: bms/go/go-gt/heap_update/heap_update.go
package test

var ()

const ()

func (h Heap) update(p []int64, i int64, G *Matrix) {
	var j int64
	for j = 0; j < G.N; j++ {
		if G.Get(i, j) > 0 {
			if h.W[i]+G.Get(i, j) < h.W[j] {
				p[j] = i + 1
				h.W[j] = h.W[i] + G.Get(i, j)
				h.up(h.A[j])
			}
		}
	}
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

func (h Heap) up(j int64) {
	for {
		i := (j - 1) / 2
		if i == j || h.less(i, j) {
			break
		}
		h.swap(i, j)
		j = i
	}
}

func (h Heap) swap(a, b int64) {
	i, j := h.I[a], h.I[b]
	h.I[a], h.I[b] = h.I[b], h.I[a]
	h.A[i], h.A[j] = b, a
}

func (h Heap) less(a, b int64) bool {
	i, j := h.I[a], h.I[b]
	return h.W[i] < h.W[j]
}

type Heap struct {
	N int64
	I []int64
	A []int64
	W []int64
}

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/fail/go/go-gt/heap_update/mistral/Hinted/heap_update.rs


use std::ops::Index;

struct Heap {
    n: i64,
    i: Vec<usize>,
    a: Vec<i64>,
    w: Vec<i64>,
}

struct Matrix {
    n: i64,
    a: Box<[i64]>,
}

impl Index<usize> for Matrix {
    type Output = i64;
    fn index(&self, index: usize) -> &Self::Output {
        &self.a[index]
    }
}

fn up(heap: &mut Heap, j: usize) {
    let mut i_j = heap.i[j];
    let mut i_i = heap.i[j];
    let mut i_parent = (j - 1) / 2;
    while i_parent != j {
        let i_parent_elem = heap.i[i_parent];
        if heap.less(i_parent_elem, i_i) {
            heap.i[j] = i_parent_elem;
            heap.i[i_parent] = i_j;
            i_i = i_parent_elem;
            i_j = heap.i[j];
            i_parent = (j - 1) / 2;
        } else {
            break;
        }
    }
    heap.i[j] = i_i;
}

impl Heap {
    fn update(&mut self, p: &mut [i64], i: i64, g: &Matrix) {
        let mut j = 0;
        for _ in 0..g.n {
            if g[j * g.n as usize + i as usize] > 0 {
                if self.w[i as usize] + g[j * g.n as usize + i as usize] < self.w[j as usize] {
                    p[j as usize] = i + 1;
                    self.w[j as usize] = self.w[i as usize] + g[j * g.n as usize + i as usize];
                    up(self, j);
                }
            }
            j += 1;
        }
    }

    fn swap(&mut self, a: usize, b: usize) {
        let temp = self.a[self.i[a]];
        self.a[self.i[a]] = self.a[self.i[b]];
        self.a[self.i[b]] = temp;
        let temp_w = self.w[a];
        self.w[a] = self.w[b];
        self.w[b] = temp_w;
    }

    fn less(&self, a: usize, b: usize) -> bool {
        self.w[a] < self.w[b]
    }
}

fn main() {}

